File: src/lib/quant/executionCosts.ts
export interface ExecutionCostConfig {
  baseSlippageBps: number;
  marketImpactBpsPer1pctAdv: number;
  maxMarketImpactBps: number;
  commissionPerShare: number;
  fixedFee: number;
}

export const DEFAULT_EXECUTION_COST_CONFIG: ExecutionCostConfig = {
  baseSlippageBps: 5.0,
  marketImpactBpsPer1pctAdv: 2.0,
  maxMarketImpactBps: 50.0,
  commissionPerShare: 0.0,
  fixedFee: 0.0,
};

export interface ExecutionCostEstimate {
  symbol: string;
  side: "buy" | "sell";
  quantity: number;
  referencePrice: number;
  grossNotional: number;
  participationRatePct: number | null;
  baseSlippageBps: number;
  marketImpactBps: number;
  effectiveSlippageBps: number;
  estimatedSlippage: number;
  commission: number;
  fixedFee: number;
  totalCost: number;
  costAsFractionOfNotional: number;
}

export interface ExecutionCostInput {
  symbol: string;
  side: "buy" | "sell";
  quantity: number;
  referencePrice: number;
  averageDailyVolume?: number | null;
  config?: Partial<ExecutionCostConfig>;
}

export function estimateExecutionCost(
  input: ExecutionCostInput,
): ExecutionCostEstimate {
  const config: ExecutionCostConfig = {
    ...DEFAULT_EXECUTION_COST_CONFIG,
    ...(input.config ?? {}),
  };
  const grossNotional = input.referencePrice * input.quantity;

  let participationRatePct: number | null = null;
  let marketImpactBps = 0;
  if (input.averageDailyVolume && input.averageDailyVolume > 0) {
    participationRatePct = (input.quantity / input.averageDailyVolume) * 100;
    marketImpactBps = Math.min(
      config.maxMarketImpactBps,
      participationRatePct * config.marketImpactBpsPer1pctAdv,
    );
  }

  const effectiveSlippageBps = config.baseSlippageBps + marketImpactBps;
  const estimatedSlippage = (grossNotional * effectiveSlippageBps) / 10_000;
  const commission = input.quantity * config.commissionPerShare;
  const fixedFee = config.fixedFee;
  const totalCost = estimatedSlippage + commission + fixedFee;

  return {
    symbol: input.symbol.toUpperCase(),
    side: input.side,
    quantity: input.quantity,
    referencePrice: round(input.referencePrice, 6),
    grossNotional: round(grossNotional, 2),
    participationRatePct:
      participationRatePct === null ? null : round(participationRatePct, 4),
    baseSlippageBps: round(config.baseSlippageBps, 4),
    marketImpactBps: round(marketImpactBps, 4),
    effectiveSlippageBps: round(effectiveSlippageBps, 4),
    estimatedSlippage: round(estimatedSlippage, 2),
    commission: round(commission, 2),
    fixedFee: round(fixedFee, 2),
    totalCost: round(totalCost, 2),
    costAsFractionOfNotional:
      grossNotional > 0 ? round(totalCost / grossNotional, 6) : 0,
  };
}

function round(x: number, digits = 2): number {
  const f = 10 ** digits;
  return Math.round(x * f) / f;
}

File: backend/app/quant/execution_costs.py
"""Conservative execution-cost estimates for pre-trade risk checks."""
from app.schemas.quant import ExecutionCostEstimate, ExecutionCostRequest


def estimate_execution_cost(request: ExecutionCostRequest, settings) -> ExecutionCostEstimate:
    """Estimate spread/slippage, impact, commission, and all-in cash impact.

    The estimate is deliberately conservative and is not a substitute for the
    broker's final fill report. Market impact is based on order participation
    when average daily volume is available and is capped by configuration.
    """
    gross_notional = request.reference_price * request.quantity
    participation_rate_pct = None
    market_impact_bps = 0.0
    if request.average_daily_volume:
        participation_rate_pct = request.quantity / request.average_daily_volume * 100.0
        market_impact_bps = min(
            settings.execution_max_market_impact_bps,
            participation_rate_pct * settings.execution_market_impact_bps_per_1pct_adv,
        )

    effective_slippage_bps = settings.execution_base_slippage_bps + market_impact_bps
    estimated_slippage = gross_notional * effective_slippage_bps / 10_000
    commission = request.quantity * settings.execution_commission_per_share
    fixed_fee = settings.execution_fixed_fee
    total_cost = estimated_slippage + commission + fixed_fee

    return ExecutionCostEstimate(
        symbol=request.symbol,
        side=request.side,
        order_type=request.order_type,
        quantity=request.quantity,
        reference_price=round(request.reference_price, 6),
        gross_notional=round(gross_notional, 2),
        participation_rate_pct=round(participation_rate_pct, 4) if participation_rate_pct is not None else None,
        base_slippage_bps=round(settings.execution_base_slippage_bps, 4),
        market_impact_bps=round(market_impact_bps, 4),
        effective_slippage_bps=round(effective_slippage_bps, 4),
        estimated_slippage=round(estimated_slippage, 2),
        commission=round(commission, 2),
        fixed_fee=round(fixed_fee, 2),
        total_cost=round(total_cost, 2),
        buy_cash_required=round(gross_notional + total_cost, 2),
        sell_net_proceeds=round(max(gross_notional - total_cost, 0.0), 2),
    )
